sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,731 @@
+# Explicit SIMD Programming Extension for DPC++
+
+## Introduction
+
+The main motivation for introducing the Explicit SIMD Programming (ESP) DPC++
+extension is enabling low-level efficient programming for Intel Gen
+architectures. More specifically, Explicit SIMD provides the following
+additional features:
+- Manual vectorization of device code using the `simd` class mapped to Gen's
+  general register file. This allows to write efficient code not relying on
+  further widening by the compiler, as with traditional SIMT programming.
+- Low-level APIs efficiently mapped to Gen architecture, such as block reads.
+
+One of the future directions is enabling this extension for other architectures,
+such as x86, with extracting and clearly marking generic and target-dependent
+API portions.
+
+## Explicit SIMD execution model
+
+Explicit SIMD execution model is a basically an equivalent of the base SYCL
+execution model with subgroup size restricted to 1. Which means in ESP
+it is guaranteed that there is a single work item in any workgroup, and it maps
+to a single hardware thread. All standard SYCL APIs continue to work,
+including `sycl::intel::sub_group` ones, which become either a no-op or
+trivial. E.g. a barrier becomes just a memory fence for a compiler, collectives
+just return the value in the single work-item. Another consequences of the unit
+subgroup size is guaranteed independent forward progress between work-items on
+some Gen architecture generations.
+
+## Explicit SIMD extension APIs
+
+Explicit SIMD APIs can be used only in code to be executed on Intel Gen
+architecture devices and the host device for now. Attempt to run such code on
+other devices will result in error. 
+
+Kernels and `SYCL_EXTERNAL` functions using ESP must be explicitly marked with
+the `EXPLICIT_SIMD` attribute macro. Subgroup size quiery within such
+functions will always return `1`.
+
+*Functor kernel*
+```cpp
+using AccTy = cl::sycl::accessor<int, 1, cl::sycl::access::mode::read_write,
+                                 cl::sycl::access::target::global_buffer>;
+class Functor {
+public:
+  Functor(int X, AccTy &Acc) : X(X), Acc(Acc) {}
+
+  EXPLICIT_SIMD void operator()() { Acc[0] += X; }
+
+private:
+  int X;
+  AccTy Acc;
+};
+```
+
+*Lambda kernel and function*
+```cpp
+EXPLICIT_SIMD SYCL_EXTERNAL
+void sycl_device_f(cl::sycl::global_ptr<int> ptr, sycl::intel::gpu::simd<float, 8> X) {
+  sycl::intel::gpu::flat_block_write(*ptr.get(), X);
+}
+...
+  Q.submit([&](cl::sycl::handler &Cgh) {
+    auto Acc1 = Buf1.get_access<cl::sycl::access::mode::read>(Cgh);
+    auto Acc2 = Buf2.get_access<cl::sycl::access::mode::read_write>(Cgh);
+
+    Cgh.single_task<class KernelID>([=] () EXPLICIT_SIMD {
```

Comment:
Why a single_task with a macro at the end and not a different handler entry? `Cgh.explicit_simd`  ?

Reply:
In our original proposal, there were new invocation APIs for explicit SIMD kernels - such as `parallel_for_task`, but after internal discussion it was decided to go with std SYCL APIs for invocation. From the technical standpoint, it is nice to have fewer APIs too.

Reply:
Well we could have a single "parallel_for" for everything but sometimes is better to distinguish entry points, specially if the programming model is very different. That macro in the lambda is easy to miss, specially if the kernel lambda is defined elsewhere.
Do you expect at some point that `single_task` without any attribute would be able to use this explicit simd style?
In the overall OneAPI / DPCPP / SYCL world, I am a bit concerned that people use this explicit SIMD thinking is just part of the standard, but it is easy to miss it is a (perfectly valid) vendor extension.

Reply:
Yet another typical use for kernel properties, for example based on https://www.youtube.com/watch?v=Fp8DuVWesT4 and https://gitlab.khronos.org/sycl/Specification/issues/296

I have an overdose of macros and user-visible attributes. :-)

Reply:
The plan is to eventually have `h.single_task<...>([=]() [[intel::reqd_sub_group_size(1)]] {...})` define the explicit SIMD kernel w/o any extra compiler options, as SIMD kernel can be thought of as subgroup size 1 kernel. @rolandschulz actually suggested to replace EXPLICIT_SIMD with [[sycl_explicit_simd]] and then deprecate it once the above plan is supported.
So from that standpoint, standard SYCL invocation API seems better choice.

Reply:
This is a vendor extension, so clearly is your call. 
One of my concerns is that you could have the kernel separated from the single_task point of entry, and you may not even know what kind of kernel is that you are dispatching, just a SYCL kernel. Without a way for a C++ developer to identify the usage of this different API inside, they will be reduced to random runtime errors if the wrong kernel is used in the wrong device.

Reply:
You are right, we already experience problems with distinguishing ESIMD kernels from others. Recently we discussed another approach internally, which is closer to parallel_for_task - use extra template parameter for parallel_for which will specify some "executor", which would be better aligned with modern trends.

Reply:
added TODO to the bottom of the doc